File: Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    /**
     * @param s: a string
     * @return: an integer 
     */
    int lengthOfLongestSubstring(string s) {
        // write your code here
        unordered_map<char, size_t> last_occur;
        size_t startId = 0, ret = 0;
        for (size_t i = 0; i < s.length(); ++i)
        {
            if (last_occur.find(s[i]) == last_occur.end())
            {
                last_occur.emplace(s[i], i);
            }
            else
            {
                if (last_occur[s[i]] >= startId)
                {
                    ret = max(ret, i - startId);
                    startId = last_occur[s[i]] + 1;
                }
                last_occur[s[i]] = i;
            }
        }
        ret = max(ret, s.length() - startId);
        return ret;
    }
};

File: Longest Substring Without Repeating Characters.py
class Solution:
    # @param s: a string
    # @return: an integer
    def lengthOfLongestSubstring(self, s):
        # write your code here
        
        # length = len(s)
        # sets = set()
        # pre, post = 0, 0
        # ans = 0
        # # while pre < length:
        # for pre in range(length):
        #     if s[pre] not in sets:
        #         sets.add(s[pre])
        #         # pre += 1
        #         ans = max(ans, pre - post + 1)
        #     else:
        #         # ans = max(ans, len(sets))
        #         while s[post] != s[pre]:
        #             sets.remove(s[post])
        #             post += 1
        #         # sets.remove(s[post])
        #         post += 1
        #         # pre += 1
                
            
        # # ans = max(ans, len(sets))
            
        # return ans
        length = len(s)
        sets = set()
        j = 0
        ans = 0
        for i in range(length):
            while j < length and s[j] not in sets:
                sets.add(s[j])
                ans = max(ans, j - i + 1)
                j += 1
            if s[i] in sets:
                sets.remove(s[i])
        return ans
                